xds/internal/xdsclient/clientimpl.go

```diff
@@ -100,3 +211,151 @@ func (c *clientImpl) close() {
+		}
+		delete(s.interestedAuthorities, callingAuthority)
+	}
+
+	return c.getOrCreateChannel(serverConfig, initLocked, deInitLocked)
+}
+
+func (c *clientImpl) getChannelForLRS(serverConfig *bootstrap.ServerConfig) (*xdsChannel, func(), error) {
+	if c.done.HasFired() {
+		return nil, nil, ErrClientClosed
+	}
+
+	initLocked := func(s *channelState) { s.lrsRefs++ }
+	deInitLocked := func(s *channelState) { s.lrsRefs-- }
+
+	return c.getOrCreateChannel(serverConfig, initLocked, deInitLocked)
+}
+
+func (c *clientImpl) getOrCreateChannel(serverConfig *bootstrap.ServerConfig, initLocked, deInitLocked func(*channelState)) (*xdsChannel, func(), error) {
+	c.channelsMu.Lock()
+	defer c.channelsMu.Unlock()
+
+	if c.logger.V(2) {
+		c.logger.Infof("Received request for a reference to an xdsChannel for server config %q", serverConfig)
+	}
+
+	// Use an active channel, if one exists for this server config.
+	if state, ok := c.xdsActiveChannels[serverConfig.String()]; ok {
+		if c.logger.V(2) {
+			c.logger.Infof("Reusing an active xdsChannel for server config %q", serverConfig)
+		}
+		initLocked(state)
+		return state.channel, c.releaseChannel(serverConfig, state, deInitLocked), nil
+	}
+
+	// If an idle channel exists for this server config, remove it from the
+	// idle cache and add it to the map of active channels, add return it.
+	if state, ok := c.xdsIdleChannels.Remove(serverConfig.String()); ok {
+		if c.logger.V(2) {
+			c.logger.Infof("Reviving an xdsChannel from the idle cache for server config %q", serverConfig)
+		}
+		state, _ := state.(*channelState)
+		c.xdsActiveChannels[serverConfig.String()] = state
+		initLocked(state)
+		return state.channel, c.releaseChannel(serverConfig, state, deInitLocked), nil
+	}
+
+	if c.logger.V(2) {
+		c.logger.Infof("Creating a new xdsChannel for server config %q", serverConfig)
+	}
+
+	// Create a new transport and create a new xdsChannel, and add it to the
+	// map of xdsChannels.
+	tr, err := c.transportBuilder.Build(transport.BuildOptions{ServerConfig: serverConfig})
+	if err != nil {
+		return nil, func() {}, fmt.Errorf("failed to create transport for server config %s: %v", serverConfig, err)
+	}
+	state := &channelState{
+		parent:                c,
+		serverConfig:          serverConfig,
+		interestedAuthorities: make(map[*authority]bool),
+	}
+	channel, err := newXDSChannel(xdsChannelOpts{
+		transport:          tr,
+		serverConfig:       serverConfig,
+		bootstrapConfig:    c.config,
+		resourceTypeGetter: c.resourceTypes.get,
+		eventHandler:       state,
+		backoff:            c.backoff,
+		watchExpiryTimeout: c.watchExpiryTimeout,
+		logPrefix:          clientPrefix(c),
+	})
+	if err != nil {
+		return nil, func() {}, fmt.Errorf("failed to create xdsChannel for server config %s: %v", serverConfig, err)
+	}
+	state.channel = channel
+	c.xdsActiveChannels[serverConfig.String()] = state
```

Comment:
same here. why 304 and 305 don't need to be within critical section?

Reply:
I grab the lock at the top of this method. Are you saying that is not enough?